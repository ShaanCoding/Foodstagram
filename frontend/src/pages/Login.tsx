import styles from '../styles/Login.module.css'
import Foostaram from '../images/Foostaram.svg'
import InputField from '../components/form/InputField'
import SubmitButton from '../components/form/SubmitButton'
import Form from '../components/form/Form'
import UseLoginMutation from '../api/UseLoginMutation'
import Cookies from 'js-cookie'
import { useNavigate } from 'react-router-dom'
import UseTwoFAEnabledMutation from '../api/UseTwoFAEnabledMutation'
import { AxiosError } from 'axios'

const Login = () => {
	const navigate = useNavigate()
	const loginMutation = UseLoginMutation()
	const has2FAMutation = UseTwoFAEnabledMutation()

	if (loginMutation.isSuccess) {
		const token = loginMutation?.data?.data?.accessToken
		const hasTwoFactor = loginMutation?.data?.data?.hasTwoFactor

		if (token) {
			Cookies.set('access_token', token)
			if (hasTwoFactor === false) {
				navigate('/2fa')
			} else {
				navigate('/')
			}
		}
	}

	const awaitingEmailVerification =
		loginMutation?.data?.data?.awaitingEmailVerification

	return (
		<div className="m-12 w-4/5 h-full ml-auto mr-auto flex max-w-[750px]">
			<div className="hidden md:block flex-auto w-32 mr-8">
				<img
					alt="Promotional screenshots of the website"
					src="/images/login/login-screenshot.png"
					width={380}
					height={581}
				/>
			</div>
			<div className={`flex flex-col justify-center flex-auto w-14`}>
				<div className={`pt-12 pb-6 px-8 bg-white border ${styles.greyBorder}`}>
					<img
						alt="Foodstaram Logo"
						className="h-10 ml-auto mr-auto mb-8"
						src={Foostaram}
					/>
					{loginMutation.isError && (
						<div className="my-8 bg-red-300 rounded-lg p-4 text-center">
							{loginMutation.error instanceof AxiosError &&
							loginMutation.error?.response?.status === 401
								? loginMutation.error.response.data.message
								: 'Invalid credentials, please try again'}
						</div>
					)}

					{loginMutation.isSuccess && (
						<div className="my-8 bg-green-300 rounded-lg p-4 text-center">
							{awaitingEmailVerification
								? 'Your account is not verified. Please check your email!'
								: `Welcome ${loginMutation?.data?.data?.data?.name}! You've logged in succesfully!`}
						</div>
					)}
					<Form
						onSubmit={(data) => {
							loginMutation.mutate({
								email: data['email'],
								password: data['password'],
								otp: data['otp'],
							})
						}}
					>
						<InputField
							placeholder="Email address"
							type="email"
							name="email"
							autoComplete="foostagram-email"
							onChange={(val: string) => {
								has2FAMutation.mutate({ email: val })
							}}
							required
						/>
						<InputField
							placeholder="Password"
							type="password"
							name="password"
							autoComplete="foostagram-password"
							required
						/>

						<InputField
							placeholder="2FA Token"
							type={
								has2FAMutation.isSuccess
									? has2FAMutation.data?.data.enabled
										? 'number'
										: 'hidden'
									: 'hidden'
							}
							name="otp"
							autoComplete="foostagram-otp"
						/>
						<div className="mb-6" />
						{/*<p className="text-xs text-center text-gray-400 mt-2 mb-6">
							<a className="font-medium text-gray-500" href="/">
								Forgotten your password?
							</a>
					</p>*/}
						<SubmitButton text="Log In" loading={loginMutation.isLoading} />
					</Form>
					<p className="text-sm text-center pt-3">
						<a className="text-insta-green" href="/passwordreset">
							Forgot password?
						</a>
					</p>
				</div>
				<div className={`mt-2 py-4 px-8 bg-white border ${styles.greyBorder}`}>
					<p className="text-sm text-center">
						Don't have an account?{' '}
						<a className="text-insta-green" href="/register">
							Sign Up
						</a>
					</p>
				</div>
			</div>
		</div>
	)
}

export default Login
