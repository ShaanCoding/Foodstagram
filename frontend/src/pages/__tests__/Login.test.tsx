import { QueryClient, QueryClientProvider } from 'react-query'
import { BrowserRouter, Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import Login from '../Login'

const queryClient = new QueryClient()

it('The Login page renders correctly', () => {
	const tree = renderer
		.create(
			<QueryClientProvider client={queryClient}>
				<BrowserRouter>
					<Login />
				</BrowserRouter>
			</QueryClientProvider>
		)
		.toJSON()
	expect(tree).toMatchSnapshot()
})
