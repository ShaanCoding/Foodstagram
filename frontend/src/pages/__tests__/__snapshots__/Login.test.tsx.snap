// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Login page renders correctly 1`] = `
<div
  className="m-12 w-4/5 h-full ml-auto mr-auto flex max-w-[750px]"
>
  <div
    className="hidden md:block flex-auto w-32 mr-8"
  >
    <img
      alt="Promotional screenshots of the website"
      height={581}
      src="/images/login/login-screenshot.png"
      width={380}
    />
  </div>
  <div
    className="flex flex-col justify-center flex-auto w-14"
  >
    <div
      className="py-12 px-8 bg-white border undefined"
    >
      <img
        alt="Foodstaram Logo"
        className="h-10 ml-auto mr-auto mb-8"
        src="test-file-stub"
      />
      <form
        onSubmit={[Function]}
      >
        <input
          className="p-2 w-full rounded-sm mb-2 border hover:border-gray-500 focus-visible:border text-xs bg-white undefined"
          initialValue=""
          name="email"
          onChange={[Function]}
          placeholder="Email address"
          type="email"
          value=""
        />
        <input
          className="p-2 w-full rounded-sm mb-2 border hover:border-gray-500 focus-visible:border text-xs bg-white undefined"
          initialValue=""
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
        <p
          className="text-xs text-center text-gray-400 mt-2 mb-6"
        >
          <a
            className="font-medium text-gray-500"
            href="/"
          >
            Forgotten your password?
          </a>
        </p>
        <button
          className="w-full bg-insta-green text-white text-sm p-[5px] rounded-md font-medium"
          disabled={false}
          loading={false}
          text="Log In"
          type="submit"
        >
          Log In
        </button>
      </form>
    </div>
    <div
      className="mt-2 py-4 px-8 bg-white border undefined"
    >
      <p
        className="text-sm text-center"
      >
        Don't have an account?
         
        <a
          className="text-insta-green"
          href="/register"
        >
          Sign Up
        </a>
      </p>
    </div>
  </div>
</div>
`;

exports[`renders correctly 1`] = `
<div
  className="m-12 w-4/5 h-full ml-auto mr-auto flex max-w-[750px]"
>
  <div
    className="hidden md:block flex-auto w-32 mr-8"
  >
    <img
      alt="Promotional screenshots of the website"
      height={581}
      src="/images/login/login-screenshot.png"
      width={380}
    />
  </div>
  <div
    className="flex flex-col justify-center flex-auto w-14"
  >
    <div
      className="py-12 px-8 bg-white border undefined"
    >
      <img
        alt="Foodstaram Logo"
        className="h-10 ml-auto mr-auto mb-8"
        src="test-file-stub"
      />
      <form
        onSubmit={[Function]}
      >
        <input
          className="p-2 w-full rounded-sm mb-2 border hover:border-gray-500 focus-visible:border text-xs bg-white undefined"
          initialValue=""
          name="email"
          onChange={[Function]}
          placeholder="Email address"
          type="email"
          value=""
        />
        <input
          className="p-2 w-full rounded-sm mb-2 border hover:border-gray-500 focus-visible:border text-xs bg-white undefined"
          initialValue=""
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
        <p
          className="text-xs text-center text-gray-400 mt-2 mb-6"
        >
          <a
            className="font-medium text-gray-500"
            href="/"
          >
            Forgotten your password?
          </a>
        </p>
        <button
          className="w-full bg-insta-green text-white text-sm p-[5px] rounded-md font-medium"
          disabled={false}
          loading={false}
          text="Log In"
          type="submit"
        >
          Log In
        </button>
      </form>
    </div>
    <div
      className="mt-2 py-4 px-8 bg-white border undefined"
    >
      <p
        className="text-sm text-center"
      >
        Don't have an account?
         
        <a
          className="text-insta-green"
          href="/register"
        >
          Sign Up
        </a>
      </p>
    </div>
  </div>
</div>
`;
